Tiny-Home-Consulting/DataOnion
# Feature/THOS-8: Two Factor Auth

## Issue Link

Jira issue: https://tinyhomeconsultingllc.atlassian.net/browse/THOS-8

## Overview of Changes

This PR adds two-factor authentication to DataOnion, using an implementation that closely mirrors that used in CPC Realtime API.

### Anything you want to highlight?

This PR also adds `TwoFactorRedisContext` and two `ApiException` classes, which `TwoFactorAuthService` depends on.

I changed the return types for the `IFluentRedisOnion` and `IFluentAuthStategyOnion` methods so that the fluent methods could be chained together more freely, and the client wouldn't have to refer to a specific order.

## Test Plan

I created a branch for the CPC Realtime API that allows us to test this implementation manually, and demonstrates the changes that would have to be made for a project using 2FA to switch over to DataOnion. It can be found here:

[https://github.com/combined-public-communications/cpc-realtime-api/tree/2fa-test](https://github.com/combined-public-communications/cpc-realtime-api/tree/2fa-test)

To test this implementation manually:
1. Pull down the latest commits from `feature/THOS-8-two-factor-auth` and `2fa-test`.
2. Make sure that CPC Realtime API is using the local copy of DataOnion. If the two projects are in the same directory, then in _both_ `.csproj` files in the project, the project reference for DataOnion should be:
```
<ProjectReference Include="..\..\DataOnion\DataOnion\DataOnion.csproj" />
```
3. Run the API.
4. In Postman, hit the endpoint for `Register DID to CPC User`, using a DID attached to a device you own (such as your mobile phone). The device owning that DID should receive a text or a call, depending on which option you chose.
5. Hit the endpoint again, at least once. You should receive a response with the message `"Too many requests. Please wait [number] seconds."`. The timeout should decline with each request. You should not see a stack trace.
6. Using the verification code you received on your device, hit the endpoint for `Verify DID for CPC User` on Postman. You should receive a response stating that your DID number was verified.
7. Run the unit tests. (NOTE: As of 11/8/22, 4 unit tests in CPC Realtime API are failing due to changes made in [THOS-10](https://github.com/Tiny-Home-Consulting/DataOnion/pull/10). This is not related to two-factor authentication and should be addressed in a separate PR).
